PartnerLibraryInstrumentationGuidelines.md

```diff
@@ -0,0 +1,193 @@
+# Guidelines for instrumenting partner libraries with Diagnostic Source 
+
+This document provides guidelines for adding Diagnostic Source instrumentation to external libraries in a way that will allow to automatically collect high quality telemetry in Application Insights SDK.
+
+## Diagnostic Source and Activities
+
+[Diagnostic Source][DiagnosticSourceGuide] is a simple module that allows code to be instrumented for production-time logging of rich data payloads for consumption within the process that was instrumented. At runtime consumers can dynamically discover data sources and subscribe to the ones of interest.
+
+[Activity][ActivityGuide] is a class that allows storing and accessing diagnostics context and consuming it with logging system.
+
+Both Diagnostic Source and Activity have been used to instrument [System.Net.Http][SystemNetHttp] and [Microsoft.AspNetCore.Hosting][MicrosoftAspNetCoreHosting], although that instrumentation is not fully complaiant with this guidance.
+
+[This document][DiagnosticSourceActivityHowto] can help to a better understanding on how to efficiently use Diagnostic Source.
+
+### Instrumentation code
+
+The following code sample shows how to instrument the operation logic enclosed in ```ProcessOperationImplAsync()``` method, in the most efficient way which will ensure no performance overhead is added if there are no listeners for that particular Activity.
+
```

Comment:
Thanks Tomasz for soliciting my feedback.      What makes logging 'hard' is the distributed nature of the instrumentation points (being authored by many people) which makes consistency hard and pretty much all problem grow from there.  

The ideal situation is if the guidance is 'easy' and hard to mess up.  As you can see, this is not the case for the sample code below (We are suggesting ~100 lines of code JUST for instrumenting ONE start-stop pair.   WE CAN'T GO OUT WITH THIS AS THE GUIDANCE!   It just won't get done correctly/uniformly.  

First, I would like to break the guidance up, and first show the synchronous case (where the operation represented by the activity is simply a call to a function on the same thread and not an async Task)  Then we can show the non-async case (which is a bit more complex, but that is Async complexity not logging complexity).  

But fundamentally we need to cut the lines of code down.   If we take the synchronous case, I really am trying to get as close to the following code as possible (which is still more complex than I want, but not too bad).  Note that I am assuming APIs that don't currently exist but which I would like to exist.   
```
    Activity activity = null;
    if (source.IsEnabled(ActivityName)) // This is here for efficiency (avoiding creating payload object when off.  
        activity = source.StartActivity(ActivityName, new {Input = input });

    var output DoOperation(input); // Do the actual operation (probably more than just one function call) 

    if (activity != null)
         source.StopActivity(activity,  new  { Output = output }); 
```
Of cource I need my new overloads for some methods, but that is trivial and I will ignore it for now...
The main differences between what you have and what I want are the following
1. I don't bother with exception cases or logging the exceptions.
2. I don't add any tags to the activity. (you have component, span.kind, error).   

Note that I would not have a problem with particular important components adding these additional things (like the HTTP incoming and outgoing activities), but I don't really want to mandate that every component should be doing this.   

The question is how bad is it that I have eliminated these things in the common case?     Certainly I am open to negotiation here, but I would argue that it is not bad.   Exceptions are already heavily logged, and they naturally get stamped with the correct activity, so it is not clear we need per-component processing of exceptions.    The issue of the tags are more interesting.  'error' feels redundant with the exception being thrown, so I would not want to mandate it unless we were forced to (especially if we have it on the 'boundary' where HTTP requests come into the system).   span.kind = client is certinaly something that the infrastructure could do for you (it is a constant for all code here correct?.   Component is interesting, but feels like it overlaps with the name of the DiagnosticSource.  I would like to understand the real value here.   It seems like we could live without it (at least for most components). 

This probably needs more discussion, but this is the 'big rock'. 

As an aside, the code we show should definitely a true sample (where we just have  Console.WriteLine for the component body.    When looking at the code below I *think* I see errors that would cause compilation failure.   I don't want to dwell on those until we agree with the actual code that should be in the guidance.

Once again, I want to thank you letting me review.   We are trying to make the logging experience as good as it can be and I think cutting the complexity of what we ask people to write at an instrumentation point is a key aspect of 'getting it right'.   It is worth our time...

Reply:
Thanks @vancem - as mentioned offline, I agree with some of the things you brought up (particularly complexity of the code sample) and will address those (in separate PR).